src/annoylib.h

```diff
@@ -137,6 +167,108 @@ inline void two_means(const vector<Node*>& nodes, int f, Random& random, bool co
   }
 }
 
+#ifdef USE_AVX
+static inline float get_norm(float *v, int f) {
+  float sq_norm = 0;
+  int i = f;
+  if (f > 7) {
+    __m256 norm = _mm256_setzero_ps();
+    for (; i > 7; i -= 8) {
+      const __m256 v_v = _mm256_loadu_ps(v);
+      norm = _mm256_add_ps(norm, _mm256_mul_ps(v_v, v_v));
+      v += 8;
+    }
+    // Sum all floats in norm register.
+    sq_norm = hsum256_ps_avx(norm);
+  }
+  // Don't forget the remaining values.
+  for (; i > 0; i--) {
+    sq_norm += *v * *v;
+    v++;
+  }
+  return sqrt(sq_norm);
+}
+
+static inline void normalize(float *v, int f) {
+  float norm = get_norm(v, f);
+  __m256 v_norm = _mm256_set1_ps(norm);
+
+  int i = f;
+  for (; i > 7; i -= 8) {
+    _mm256_storeu_ps(v, _mm256_div_ps(_mm256_loadu_ps(v), v_norm));
+    v += 8;
+  }
+  // Don't forget the remaining values.
+  for (; i > 0; i--) {
+    *v /= norm;
+    v++;
+  }
+}
+
+template<typename T, typename Random, typename Distance, typename Node>
```

Comment:
why is this functino actually needed? Shouldn't the distance function already use AVX when possible?

Reply:
The distance function will use the optimized code, but I also did some optimizations in the inner for loops. Without them I was not able to measure any improvements when normalize and get_norm where also optimized.